### PR TITLE
Simplify DataStorm callback executor

### PR DESCRIPTION
**Pull request by pepone**
This PR simplifies the callback executor used by DataStorm. Specifically there is no need to queue DataElement alongs the callback, the callbacks can capture the element if required.

### CODE REVIEW

File: cpp/src/DataStorm/CallbackExecutor.cpp [hunk 1]
```diff
@@ -31,18 +31,18 @@ CallbackExecutor::CallbackExecutor(function<void(function<void()> call)> callbac
...
-                for (const auto& p : queue)
+                for (const auto& callback : queue)
                 {
                     try
                     {
                         if (_callbackExecutor)
                         {
-                            // TODO do we need to ensure p.second is executed before we continue?
-                            _callbackExecutor(p.second);
+                            // TODO do we need to ensure that the callback is executed before we continue?
```
Comment (bernardnormier) [inline cpp/src/DataStorm/CallbackExecutor.cpp:40]: I don't think we do and I don't see how we would do it.

File: cpp/src/DataStorm/CallbackExecutor.cpp [hunk 2]
```diff
@@ -31,18 +31,18 @@ CallbackExecutor::CallbackExecutor(function<void(function<void()> call)> callbac
...
+                            _callbackExecutor(callback);
```
Comment (bernardnormier) [inline cpp/src/DataStorm/CallbackExecutor.cpp:41]: I would prefer a clearer name for this field.

**Review by bernardnormier — APPROVED** [review]

File: cpp/src/DataStorm/CallbackExecutor.cpp (hunk 1 above)
Comment (pepone) [reply cpp/src/DataStorm/CallbackExecutor.cpp:40]: > I don't think we do

I guess destroy just cares about joining the thread; it just so happens that, without the custom executors, this also means all callbacks have finished executing.

>  I don't see how we would do it.

Wrapping the callback in another callback, the wrapper sets a condition variable, and the thread sending callbacks to the custom executor waits for the condition.

File: cpp/src/DataStorm/CallbackExecutor.cpp (hunk 2 above)
Comment (pepone) [reply cpp/src/DataStorm/CallbackExecutor.cpp]: Renamed to `_customExecutor`

**Pull request merged**